### DISCUSSION TITLE
Replacement for `toml` as config language

### DISCUSSION THREAD

**Discussion by VuiMuich** [leftwm/leftwm#496] [Ideas]
At various opportunities the notion sparked that with `toml` as our config language we are hitting the limitations of this language and our ideas/wishes for config features.

With #292 being meant to break config compatibility anyways we would also have a great opportunity to switch to a new language that offers more possibilities.

Let's use this discussion to evaluate possible candidates as well as track our demands on the feature set of the desired language.

Desired Features (will keep this updated with the ongoing discussion):
- support serde
- reasonable amount of crate dependencies
- not indentation sensitive
- allow for enums of different types
- some basic logic to densen repetitive config (i.e. MoveWindowToTag, GoToTag, similar keybinds with just different modiier set etc.)
- good human redability, error visibility

**Comment by AethanFoot** [comment]
Looking at the big ones, I feel yaml is the best. It quite readable and writable (if you are good with your indents. From the [serde-yaml](https://github.com/dtolnay/serde-yaml) page it also seems to address the limit of serializing enums that have variants with values that is mentioned #472. Many users may also be familiar with this format from alacritty.

Thanks

**Reply by mautamu** [reply]
I tend to agree with AethanFoot, it's familiar enough to be useful quickly. Not a huge fan of the indents, but if that's the only major downside we can come up with that's fine. 

What are all the issues that we are trying to resolve?

**Reply by Phate6660** [reply]
YAML is for masochists. Really, it's a nightnare.
In fact, it's a major reason that I didn't use Alacritty for a long time.
Even with valid configuration (verified by Alacritty devs/maintainers) Alacritty wouldn't accept my configs.
Even to this day, I have a barely configured Alacritty.
TOML is a major improvement over it.
Personally, I think it would be a lot better if LeftWM stuck to TOML.

**Reply by AethanFoot** [reply]
I do like toml, but as pointed out @cecton it causes situations like [command.rs](https://github.com/leftwm/leftwm/blob/master/src/command.rs) where lengthy enums are required for serialization and deserialization to work (from their github "Because TOML is explicitly intended as a configuration file format, parsing it is easy, but it is not intended for serializing arbitrary data structures."). Which adds unneeded complexity to code making it harder for new contributors. 

To be fair I feel we would need to add code to leftwm-check that checks config errors related to yaml.
I do prefer the syntax of toml, but it is restrictive to more complex data types.

**Reply by hertg** [reply]
I can't fully comprehend the experience @Phate6660 had.
Since I started working with kubernetes and ansible, I've been working with YAML on a daily basis and never had any problems with it at all. I also configured alacritty without ever having any YAML related issues.
I could see that one might dislike the indentation markup though, but other than that, YAML seems fine to me.

I'd also be in favor of using a more common config language (like YAML) as a replacement for TOML rather than something exotic. It's easy to find opponents of widespread formats and then misinterpret that as the exotic option being superior when it really just comes down to a lack of users for the latter.

**Reply by cecton** [reply]
@hertg TOML is very familiar to Rust developers as this is the format of Cargo.

In this case TOML prevents validation on the parser itself. If we move away from TOML (and not necessarily to YAML) we would gain (for free):
- validation for base commands (not custom commands)
- validation for key bindings (I intend to add a type for those in a future PR)
- validation for types like `Size`
- ...maybe other types I'm not thinking of but in general: leftwm-check will have less code

About leftwm's config: it is possible to keep TOML and move a lot of the validation to the parser but it requires to delegate the parsing to other types. It's like a second layer of parsing: you want to get to `ObjectA`, then the parser would go through `RawObjectA` first and then convert the type ([serde's "try_from" attribute](https://serde.rs/container-attrs.html)). (The cost is still on leftwm-check though.)

**Reply by cecton** [reply]
I forgot to mention it but leftwm's config has very little nesting so the issue with YAML's indentation is actually not dreadful.

**Reply by hertg** [reply]
@cecton Sorry for the misunderstanding, I didn't mean TOML when I wrote about exotic formats. I just wanted to promote some cautiousness about formats, that aren't known to anyone, which may come up **as a replacement for TOML**. :)

**Comment by VuiMuich** [comment]
As mentioned otherplaces [serde_dhall](https://crates.io/crates/serde_dhall/) is a really interesting language and ticks quite a bunch of desires, but comes with a ton of dependencies (iirc at least about 50 crates).

Regarding keybinds basically the sequencing feature of sxhkd is what I'd like to have in our config-language.

**Comment by cecton** [comment]
@VuiMuich I would like to add to the discussion that TOML is currently used not only for the config file but also for the state file and we **can** have JSON for the state and TOML for the config if we want.

Using JSON for the state is helpful because we can serialize/deserialize any kind of struct/enum (which is my main complain about TOML).

**Comment by cecton** [comment]
I think there are 2 point of views on this discussion:
- user perspective
- developer perspective

### User perspective

The user will want an easy/human format:
- TOML is good: comments are nice, easy to write
- TOML is bad for specific things: for nested and graph-like structures (not the case of leftwm at the moment)
- JSON is bad: you can't add comments, you can't add `,` at the final element of arrays and objects (but good for nesting)
- JSON5 and Jsonnet are good: you can add comments, you can add `,` at the final element of arrays and objects (also good for nesting)
- YAML is "good" but losing popularity because of it's annoying indentation issue (also good for nesting)
- a whole lot of good new formats exist but people's resistance to change will make their adoption slow: [Dhall](https://dhall-lang.org/), [Jsonnet](https://jsonnet.org/), [JSON5](https://json5.org/), [HOCON](https://docs.spongepowered.org/stable/en/server/getting-started/configuration/hocon.html), [kdl](https://github.com/kdl-org/kdl)

### Developer perspective

The developer wants to leverage the power of the serializer/deserializer:
- TOML is bad: tuple-like enum variants and struct-like enum variants are forbidden
- JSON and YAML can serialize and deserialize everything, so they're equally good
- JSON5, ... others ... are equally good as long as it is supported by serde

### Both perspectives

Now if I take a very "large" point of view:
- Because leftwm doesn't use much nesting in its config file, we could maintain the code that will convert the "human" TOML to the final structs. It's a cost though.
- Using JSON5 or something else seem to be the best solution. The only issue is that some people will complain that they are "new"/"not standard" formats.
- YAML is standard but will make some people unhappy because of its annoying indentation.
- JSON is standard but not human-friendly.

### Possible solutions

From there I see 3 paths:
1. We choose TOML:
    - no other format will be allowed
    - it's human-friendly alright
    - we need to maintain a bit of code for conversion
2. We choose one of the other formats (YAML, JSON5, etc...):
    - no other format will be allowed
    - it's human-friendly if we don't choose JSON
    - no extra code to maintain
    - it's "standard" if we choose YAML
3. We choose all the formats that are serializable through serde and can serialize our structs:
    - this will exclude TOML
    - this will include: JSON, YAML, JSON5, ... probably a bunch of others supported with serde
    - it's extendable: people can submit PRs to add a new format (if the serde impl exists)
    - it's as human-friendly as the user wants (if they prefer JSON over YAML or YAML over JSON they will have the choice)
    - no extra code to maintain (it's delegated to the serializer implementations so it's no code for us)
    - more dependencies depending on what format we add **BUT** we can put them behind feature flags

### My conclusion

To me solution 1 and 2 are equal as they come with their pros and cons. I personally prefer the solution 3 because it's more fancy and I'm a fancy person :grin: but also because it gives more choice to the user. People who want to optimize for size will be able to do it using the feature flags. Everybody will choose their format of choice... it would be best if the default flags would include all formats though (to reach more people).

Now, that was my personal preference, but I think the "wise" and simplest solution is actually YAML because it has the most "pros": human-friendly, less dependencies, no extra code.

**Reply by cecton** [reply]
- Hocon & serde: https://crates.io/crates/hocon (seems very good)
- Jsonnet & serde: couldn't find one
- JSON5 & serde: https://crates.io/crates/json5 (actively maintained and popular)
- kdl & serde: https://crates.io/crates/serde_kdl/0.1.0 (seems very new)
- Dhall & serde: https://crates.io/crates/serde_dhall (seems very good but lot of dependencies)

**Reply by AethanFoot** [reply]
I enjoy the idea of option 3, however in the release (or no flags) we should default to a main one, say YAML. Then the user can change the format with a flag (where we could setup multiple versions on the aur and maybe we could setup some pre-compiled release binaries).

**Reply by mautamu** [reply]
Hmm, will having multiple diverging configuration methods make debugging more complex for bug triage?

There will also be the issue of themes, theme A uses YAML, B uses JSON5, and the user wants both themes to work without having to recompile leftwm — though I suppose we could wash our hands of this and say that’s a downstream problem, it might be helpful to either have a standard for themes, to enable all by default, or to ask leftwm-theme to track what config file formats leftwm supports against what each theme supports.

**Reply by VuiMuich** [reply]
Just skimmed over the reamdme pages of `HOCON` and `kdl`, I have to say especially the latter really intrigues me.
Plus from a brief look in ther issue tracker they seem to be some lovely folks.

In the 'adventurer spirit' of leftwm I'd say switch to `kdl` with the 'official' builds and to keep leftwm-community-themes maintainable.
Additionally we could have a guide in the wiki to enable the most adventurous users to add support for a config language of their own liking via build-flags.

EDIT: Also `kdl` looks very lean on dependencies 👍 and maybe a little conversion tool to translate toml to kdl should not be too complicated

**Reply by cecton** [reply]
+1 for the adventurer spirit :grin:

**Reply by hertg** [reply]
Wow, didn't think to see HOCON here. I think you linked a readme that is a bit limited, the [original](https://github.com/lightbend/config/blob/master/HOCON.md) is much more thorough. 

My two cents as a long-time user of HOCON: 
I've always liked HOCON a lot and have used it since years because I was working with Play! Framework. Can't really say anything bad about it, it's essentially JSON without the annoyances of writing json. You are also allowed to write plain json inside a HOCON file, I don't know why you would though :stuck_out_tongue_closed_eyes: 

I actually thought about mentioning it yesterday, but felt that it was a bit too niche. I've never encountered HOCON anywhere else than in Lightbend products like Play! and Akka.

**Reply by cecton** [reply]
ohhhh kdl2? 😋  https://twitter.com/zkat__/status/1445647664826093578?s=20

**Reply by cecton** [reply]
RON is pretty cool https://github.com/ron-rs/ron

**Comment by di-effe** [comment]
Hi, not a developer so I am a little out of my league here, but being a sysadmin for many years I can assure you that YAML will generate a lot of frustration for any average Joe who will try and discard quickly TileWM.

**Reply by cecton** [reply]
Thank you for the feedback :heart:

**Reply by hertg** [reply]
Can you elaborate what kind of frustrations to expect? I'm forced to use YAML a lot, but I never had any issues with it (from a user-only perspective). I came to realize that I'm the only one with that experience though, so I'm genuinely interested.

I'm not an advocate for YAML in particular, but I usually prefer when projects use config languages I'm familiar with rather than having to learn a new config language with every project I use :stuck_out_tongue_closed_eyes: 

That being said, I'm pretty open-minded on what config language LeftWM should use eventually and I'll be happy to learn and adept.

**Reply by di-effe** [reply]
I got nothing personal against YAML, I use it daily for k8s :)
But you have to admit that it is easy to break it. If you know why it's also easy to fix. If you don´t, you simply conclude that it is buggy.

In the end I guess my point is: if you (not you YOU of course) would like to lure newcomers, probably not tech-savvy, you'll have to face the fact that they **will** destroy a YAML file the first time they try to change something .

That said, how realistic it is that a tiling vm would become mainstream? I don´t know, probably very low. 
It was just a thought :)
